crates/noirc_evaluator/src/ssa/opt/die.rs

```diff
@@ -92,12 +92,26 @@ impl Context {
 
         let instruction = &function.dfg[instruction_id];
 
-        // These instruction types cannot be removed
-        if matches!(
-            instruction,
-            Constrain(_) | Call { .. } | Store { .. } | EnableSideEffects { .. }
-        ) {
-            return false;
+        match instruction {
+            // These instruction types can be safely removed
+            Binary(_)
+            | Cast(_, _)
+            | Not(_)
+            | Truncate { .. }
+            | Allocate
+            | Load { .. }
+            | ArrayGet { .. }
+            | ArraySet { .. } => (),
+
+            // These instruction types cannot be removed
+            Constrain(_) | Store { .. } | EnableSideEffects { .. } => return false,
+
+            Call { func, .. } => match function.dfg[*func] {
+                Value::Intrinsic(intrinsic) if !intrinsic.has_side_effects() => {
+                    // Intrinsics without side-effects can be safely removed.
+                }
+                _ => return false,
```

Comment:
```suggestion
                // Some intrinsics like println have side effects that we should not remove
                Value::Intrinsic(intrinsic) if intrinsic.has_side_effects() => {
                    return false;
                }
                _ => (),
```
This read a bit clearer to me

Reply:
This defaults to all `Value`s being removable by DIE unless it's explicitly an intrinsic which has side effects. I'm not sure that's valid for some variants of `Value`, e.g. foreign calls (once brillig gen happens after ssa optimisations) so I'm trying to keep it as constrained as possible.

Reply:
You're right, my last match case should have been `Value::Intrinsic(_) => (),`. Since that makes that approach slightly more complicated though I'll resolve this comment.